projects/dataset_recorder/dataset_recorder.cc

```diff
@@ -0,0 +1,172 @@
+// BoB robotics includes
+#include "common/background_exception_catcher.h"
+#include "common/bn055_imu.h"
+#include "common/gps.h"
+#include "common/macros.h"
+#include "common/map_coordinate.h"
+#include "common/stopwatch.h"
+#include "imgproc/opencv_unwrap_360.h"
+#include "navigation/image_database.h"
+#include "robots/rc_car_bot.h"
+#include "video/panoramic.h"
+
+// Third-party includes
+#include "plog/Log.h"
+
+// POSIX includes
+#include <unistd.h>
+
+// Standard C includes
+#include <ctime>
+
+// Standard C++ includes
+#include <chrono>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <thread>
+
+using namespace BoBRobotics;
+using namespace std::literals;
+using namespace units::angle;
+using namespace units::length;
+using namespace units::literals;
+using namespace units::time;
+
+int
+bobMain(int argc, char *argv[])
+{
+    // setting up
+    const char *path_linux = "/dev/ttyACM1"; // path for linux systems
+    Robots::RCCarBot bot;
+    GPS::Gps gps;
+    gps.connect(path_linux);
+    BN055 imu;
+    const int maxTrials = 3;
+    int numTrials = maxTrials;
+
+    const millisecond_t runTime = (argc > 1) ? second_t{ std::stod(argv[1]) } : 30_s;
+    LOGD << "running for " << runTime;
+
+    // check to see if we get valid coordinates by checking GPS quality
+    GPS::GPSData gpsData;
+    for (; numTrials > 0; numTrials--) {
+        try {
+            gpsData = gps.getGPSData();
+        } catch (GPS::GPSError &e) {
+            LOGW << " measuring failed, trying again in 1 second "
+                 << "[" << maxTrials - numTrials << "/" << maxTrials << "]";
+            std::this_thread::sleep_for(1s);
+            continue;
+        }
+
+        if (gpsData.gpsQuality != GPS::GPSQuality::INVALID) {
+            LOGI << " we have a valid measurement";
+            break;
+        }
+    }
```

Comment:
might it be worth putting this logic into either  a method in ``GPS::GPSData`` or a free function in ``GPS`` - it seems pretty standard to most of the code I've seen go by and it seems synonymous with functions like ``See3CAM_CU40::autoExposure``

Reply:
Good point

Reply:
Hmm... maybe it's better to address GPS issues in #213 ? I'd still like to merge that at some point, but it's pending testing by @tenxlenx .

Reply:
Sure